webapp/sources/rudder/rudder-core/src/main/scala/com/normation/plugins/PluginSettings.scala

```diff
@@ -273,3 +281,259 @@ class FilePluginSettingsService(pluginConfFile: File, readSetupDone: IOResult[Bo
+  }
+
+  final case class RudderAbiVersionError(rudderFullVersion: String) extends PluginManagementError {
+    override def kind:       Kind.AbiVersionError.type = Kind.AbiVersionError
+    override def displayMsg: String                    =
+      s"This plugin was not built for current Rudder ABI version ${rudderFullVersion}. You should update it to avoid code incompatibilities."
+  }
+
+  def fromRudderPackagePlugin(
+      plugin: RudderPackagePlugin
+  )(implicit rudderFullVersion: String, abiVersion: Version): List[PluginManagementError] = {
+    List(
+      validateAbiVersion(rudderFullVersion, abiVersion),
+      plugin.license.flatMap(l => validateLicenseExpiration(l.endDate))
+    ).flatten
+  }
+
+  private def validateAbiVersion(rudderFullVersion: String, abiVersion: Version): Option[RudderAbiVersionError] = {
+    if (rudderFullVersion != abiVersion.toVersionString)
+      Some(RudderAbiVersionError(rudderFullVersion))
+    else
+      None
+  }
+
+  /**
+   * license near expiration : 1 month before now.
+   */
+  private def validateLicenseExpiration(endDate: DateTime): Option[LicenseExpirationError] = {
+    if (endDate.isBeforeNow()) {
+      Some(LicenseExpiredError)
+    } else if (endDate.minusMonths(1).isBeforeNow())
+      Some(LicenseNearExpirationError)
+    else
+      None
+  }
+
+}
+
+final case class JsonPluginSystemDetails(
+    id:            String,
+    name:          String,
+    description:   String,
+    version:       String,
+    status:        JsonPluginSystemStatus,
+    statusMessage: Option[String],
+    abiVersion:    Version,
+    pluginType:    PluginType,
+    errors:        List[JsonPluginManagementError],
+    license:       Option[PluginLicenseInfo]
+)
+
+final case class JsonPluginManagementError(
+    error:   String,
+    message: String
+)
+object JsonPluginManagementError {
+  implicit val transformer: Transformer[PluginManagementError, JsonPluginManagementError] = {
+    Transformer
+      .define[PluginManagementError, JsonPluginManagementError]
+      .withFieldComputed(_.error, _.kind.entryName)
+      .withFieldComputed(_.message, _.displayMsg)
+      .buildTransformer
+  }
+}
+
+object JsonPluginSystemDetails {
+  implicit val encoderPluginSystemStatusRest: JsonEncoder[JsonPluginSystemStatus]    = JsonEncoder[String].contramap(_.entryName)
+  implicit val encoderPluginType:             JsonEncoder[PluginType]                = JsonEncoder[String].contramap(_.entryName)
+  implicit val encoderPluginManagementError:  JsonEncoder[JsonPluginManagementError] =
+    DeriveJsonEncoder.gen[JsonPluginManagementError]
+  implicit val encoderVersion:                JsonEncoder[Version]                   = JsonEncoder[String].contramap(_.toVersionString)
+  implicit val encoderPluginSystemDetails:    JsonEncoder[JsonPluginSystemDetails]   = DeriveJsonEncoder.gen[JsonPluginSystemDetails]
+}
+
+@jsonMemberNames(SnakeCase)
+final case class RudderPackagePlugin(
```

Comment:
latestVersion can also be absent, and there is a new `requires_license` field, see https://github.com/Normation/rudder/blob/2ff056ed5f1555f50ccdd526040aadbaec1d8708/relay/sources/rudder-package/src/list.rs#L25-L45 for details

Reply:
and we can definitely change the JSON format now it it makes things simpler for the webapp interaction, for which it is designed.

Reply:
Totally. I will accept that first PR so that the UI part can be done, and we will have more room to discuss what is needed